Feat/add median filter

njneeteson opened the pull request:
### Overview

Added a wrapper for the VTK filter vtkImageMedian3D for uniform data.

Resolves: #2114 

### Details

1. In pyvista/pyvista/core/filters/uniform_grid.py:

Added a method "median_smooth" to the abstract class UniformGridFilters called "median_smooth." The filter-specific method parameter "kernel_size" corresponds to the ".SetKernelSize" method for the vtkImageMedian3D, which requires 3 integers. Compared to gaussian smoothing, the method for median filtering is a bit less flexible in terms of how parameters are set, but this is also true for the underlying VTK classes.

It's possible we want a bit more flexibility, say have it be possible for the user to just give a single number for kernel size and then set the kernel size to be that given value, except set the size to 1 in dimensions where the shape of the underlying scalar data is 1. I wanted to do this originally but I couldn't think of an elegant way to detect this since self.dimension might give a shape of 2 for a flat dimension if the data is on cells, and when you do dataset[value].shape it's ravelled. We can leave it as-is or maybe someone could give me a pointer on a simple way to check the shape of the scalar data elegantly.

2. In pyvista/pyvista/_vtk.py:

Added vtkImageMedian3D to be imported along with vtkImageGaussianSmooth

akaszynski commented:
This looks great so far! Please add a unit test and verify coverage by checking it locally using `pytest-cov` and I'll be sure to review it.

njneeteson commented:
I added a simple unit just to make sure the output from median smoothing the example uniform grid was a uniform grid. I can add more unit tests that verify it does a correct median filter if necessary.

I also noticed there was no unit test for the gaussian smooth method as far as I could see anywhere so I added a similar simple unit test for that method. Again, if functionality testing unit tests are needed (e.g. specific test inpu and expected output) I can add those easily as well.

Test coverage report:

```
---------- coverage: platform darwin, python 3.7.11-final-0 ----------
Name                                           Stmts   Miss  Cover
------------------------------------------------------------------
pyvista/__init__.py                               61     20    67%
pyvista/_typing.py                                 7      0   100%
pyvista/_version.py                                2      0   100%
pyvista/_vtk.py                                   74      0   100%
pyvista/core/__init__.py                          10      0   100%
pyvista/core/common_data.py                        8      0   100%
pyvista/core/composite.py                        262      4    98%
pyvista/core/dataobject.py                       155      3    98%
pyvista/core/dataset.py                          593     25    96%
pyvista/core/datasetattributes.py                322      3    99%
pyvista/core/errors.py                            13      1    92%
pyvista/core/filters/__init__.py                  24      1    96%
pyvista/core/filters/composite.py                 43      0   100%
pyvista/core/filters/data_set.py                1068     28    97%
pyvista/core/filters/poly_data.py                590     44    93%
pyvista/core/filters/structured_grid.py           52      0   100%
pyvista/core/filters/uniform_grid.py              49      3    94%
pyvista/core/filters/unstructured_grid.py         10      0   100%
pyvista/core/grid.py                             203      7    97%
pyvista/core/imaging.py                           43      3    93%
pyvista/core/objects.py                          217     22    90%
pyvista/core/pointset.py                         727     47    94%
pyvista/core/pyvista_ndarray.py                   36      2    94%
pyvista/demos/__init__.py                          2      0   100%
pyvista/demos/demos.py                           204      4    98%
pyvista/demos/logo.py                            106     51    52%
pyvista/examples/__init__.py                       4      0   100%
pyvista/examples/downloads.py                     70     28    60%
pyvista/examples/examples.py                      81      0   100%
pyvista/examples/gltf.py                           2      0   100%
pyvista/examples/hdr.py                            1      0   100%
pyvista/ext/__init__.py                            0      0   100%
pyvista/ext/plot_directive.py                    189      5    97%
pyvista/jupyter/__init__.py                       35      5    86%
pyvista/jupyter/itkplotter.py                     87      3    97%
pyvista/jupyter/notebook.py                       86      9    90%
pyvista/jupyter/pv_ipygany.py                    111      1    99%
pyvista/jupyter/pv_pythreejs.py                  246      1    99%
pyvista/plotting/__init__.py                      22      3    86%
pyvista/plotting/_plotting.py                     66     14    79%
pyvista/plotting/axes.py                          27      0   100%
pyvista/plotting/background_renderer.py           48      0   100%
pyvista/plotting/camera.py                       141      4    97%
pyvista/plotting/charts.py                       960     33    97%
pyvista/plotting/colors.py                        49      5    90%
pyvista/plotting/export_vtkjs.py                 313     39    88%
pyvista/plotting/helpers.py                       91     12    87%
pyvista/plotting/lights.py                       231      0   100%
pyvista/plotting/mapper.py                       166     20    88%
pyvista/plotting/picking.py                      221     13    94%
pyvista/plotting/plotting.py                    1611    276    83%
pyvista/plotting/render_window_interactor.py     200      7    96%
pyvista/plotting/renderer.py                     932    107    89%
pyvista/plotting/renderers.py                    222      5    98%
pyvista/plotting/scalar_bars.py                  225      3    99%
pyvista/plotting/theme.py                          2      2     0%
pyvista/plotting/tools.py                        209     21    90%
pyvista/plotting/widgets.py                      612     23    96%
pyvista/themes.py                                778      1    99%
pyvista/utilities/__init__.py                     11      0   100%
pyvista/utilities/cell_type_helper.py             14      4    71%
pyvista/utilities/cells.py                       135      5    96%
pyvista/utilities/common.py                       20      0   100%
pyvista/utilities/errors.py                      167      8    95%
pyvista/utilities/features.py                     49     13    73%
pyvista/utilities/fileio.py                      241     55    77%
pyvista/utilities/geometric_objects.py           366     14    96%
pyvista/utilities/helpers.py                     441     59    87%
pyvista/utilities/misc.py                         22      2    91%
pyvista/utilities/parametric_objects.py          308     38    88%
pyvista/utilities/reader.py                      256    127    50%
pyvista/utilities/regression.py                   60      7    88%
pyvista/utilities/sphinx_gallery.py               21      3    86%
pyvista/utilities/transformations.py              63      0   100%
pyvista/utilities/wrappers.py                      2      0   100%
pyvista/utilities/xvfb.py                         17     13    24%
------------------------------------------------------------------
TOTAL                                          15112   1256    92%
```

akaszynski commented on tests/test_filters.py:
```diff
@@ -1394,6 +1394,15 @@ def test_extract_subset():
     #   https://gitlab.kitware.com/vtk/vtk/-/issues/17938
     assert voi.origin == voi.bounds[::2]
 
+def test_gaussian_smooth():
+    volume = examples.load_uniform()
+    volume_smooth = volume.gaussian_smooth()
+    assert isinstance(volume_smooth, pyvista.UniformGrid)
+
+def test_median_smooth():
+    volume = examples.load_uniform()
+    volume_smooth = volume.median_smooth()
+    assert isinstance(volume_smooth, pyvista.UniformGrid)
```
```suggestion

def test_gaussian_smooth():
    volume = examples.load_uniform()
    volume_smooth = volume.gaussian_smooth()
    assert isinstance(volume_smooth, pyvista.UniformGrid)


def test_median_smooth():
    volume = examples.load_uniform()
    volume_smooth = volume.median_smooth()
    assert isinstance(volume_smooth, pyvista.UniformGrid)

```

akaszynski commented:
Add in an additional line in each test to verify that setting scalars to a value works. I'm not sure why our code coverage isn't showing right now; I'll add that in a separate PR.

akaszynski commented on pyvista/core/filters/uniform_grid.py:
```diff
@@ -58,6 +58,46 @@ def gaussian_smooth(self, radius_factor=1.5, std_dev=2.,
+        else:
+            field = self.get_array_association(scalars, preference=preference)
+        alg.SetInputArrayToProcess(0, 0, 0, field.value, scalars) # args: (idx, port, connection, field, name)
+        alg.SetKernelSize(kernel_size[0],kernel_size[1],kernel_size[2])
```
```suggestion
        alg.SetKernelSize(kernel_size[0], kernel_size[1], kernel_size[2])
```

njneeteson commented:
OK I added the extra requested lines to the unit tests and re-ran the tests locally again. Let me know if anything else is needed.

MatthewFlamm commented:
> I can add more unit tests that verify it does a correct median filter if necessary.

One thought for the median filter is to create a scalar that is constant everywhere except at 1 point where it is much higher.  Then after applying the filter, the value should be constant everywhere.  Similarly for the `gaussian_smooth` filter, the value of the outlier should decrease.

This is nice for catching degradations in vtk too.

njneeteson commented:
Sure if we want to check if the underlying VTK objects are doing what we expect them to and not just test the wrapper method(s), I can add some functionality tests no problem.

MatthewFlamm commented:
> Sure if we want to check if the underlying VTK objects are doing what we expect them to and not just test the wrapper method(s), I can add some functionality tests no problem.

I don't insist on it, just a thought.

njneeteson commented:
OK - I added two more tests each for median and gaussian smooth. One test creates a uniform grid of 1s and makes sure that after smoothing, the values are the same as before. The second test creates a uniform grid of 1s with a single value of 100 near the center. The median filter outlier test wants to see that the outlier is replaced by a 1 (removed), while the gaussian filter outlier test just wants to check that the maximum value in the dataset (for the current active scalars) has decreased following gaussian smoothing.

'np.allclose' had to be used for the gaussian filter constant field check because sometimes fp arithmetic happens and gaussian smooth changes the field of all 1s to a field of (1 - 1e-16) or something like that.

akaszynski reviewed (approved):
This is great! Thanks for your work here.

koenterheegde0507 commented:
Is it still possible to use this filter as uniform_grid is deprecated and image_data only has the gaussian_smooth() attribute?

njneeteson commented:
I'm not sure what the situation is with the filter wrappers I added and the switch from `UniformGrid` to `ImageData` (there are some others that seem to be missing as well, like dilation and erosion). However as a workaround you can just use the original `vtk` objects directly:

```
from vtk import vtkImageMedian3D

median = vtkImageMedian3D()
<set parameters>
median.SetInput(img) # img is a pv.ImageData
median.Update()
img = pv.wrap(median.GetOutput())
```

MatthewFlamm commented:
This is an attribute on my version, although I tested on the main branch.  It also shows up in the documentation.  What version are you using that it doesn't show up correctly?

```py
In [7]: "median_smooth" in dir(pv.ImageData())
Out[7]: True
```

koenterheegde0507 commented:
Could this also be applied to meshes? I've been using Slicer3D, in which you can apply this filter to a mesh but I would have no idea how to implement that into pyvista/VTK

koenterheegde0507 commented:
> This is an attribute on my version, although I tested on the main branch. It also shows up in the documentation. What version are you using that it doesn't show up correctly?
> 
> ```python
> In [7]: "median_smooth" in dir(pv.ImageData())
> Out[7]: True
> ```

I did the same, and I get the same response. I just looked up the documentation online and there is no method displayed there saying median_smooth(). Using version 0.42.3

njneeteson commented:
Ah yeah that's what it is. I was wondering why I thought those filters were missing and it's because on the documentation for `pv.ImageData` they aren't listed. 

https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.ImageData.html#pyvista-imagedata

I guess they are actually there but just hard to find in the docs. I am on an older version of pyvista so hadn't done the direct check for the existence of the method(s).

njneeteson commented:
They are listed here though:

https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.imagedatafilters

MatthewFlamm commented:
The current state of the documentation indeed is confusing.  We used to have all the filters reproduced on all the `DataSet` classes, e.g. `ImageData`, `PolyData`, etc., but this increased the size and complexity of the documentation.  This issue is already tracked in #5694.
